src/opensearch.py

```diff
@@ -35,21 +35,61 @@
 logger = logging.getLogger(__name__)
 
 
+BASE_SNAP_DIR = "/var/snap/opensearch"
+SNAP_DATA = f"{BASE_SNAP_DIR}/current"
+SNAP_COMMON = f"{BASE_SNAP_DIR}/common"
+SNAP = "/snap/opensearch/current"
+
+
+OPENSEARCH_SNAP_METADATA = {
+    40: {
+        "paths": Paths(
+            home=f"{SNAP_DATA}/usr/share/opensearch",
+            conf=f"{SNAP_DATA}/etc/opensearch",
+            data=f"{SNAP_COMMON}/var/lib/opensearch",
+            logs=f"{SNAP_COMMON}/var/log/opensearch",
+            jdk=f"{SNAP}/usr/lib/jvm/java-21-openjdk-amd64",
+            tmp=f"{SNAP_COMMON}/usr/share/tmp",
+            bin=f"{SNAP}/usr/share/opensearch/bin",
+        ),
+    },
+    0: {
+        "paths": Paths(
+            home=f"{SNAP_DATA}/usr/share/opensearch",
+            conf=f"{SNAP_DATA}/etc/opensearch",
+            data=f"{SNAP_COMMON}/var/lib/opensearch",
+            logs=f"{SNAP_COMMON}/var/log/opensearch",
+            jdk=f"{SNAP}/usr/lib/jvm/java-17-openjdk-amd64",
+            tmp=f"{SNAP_COMMON}/usr/share/tmp",
+            bin=f"{SNAP}/usr/share/opensearch/bin",
+        ),
+    },
+}
+
+
 class OpenSearchSnap(OpenSearchDistribution):
     """Snap distribution of opensearch, only overrides properties and logic proper to the snap."""
 
-    _BASE_SNAP_DIR = "/var/snap/opensearch"
-    _SNAP_DATA = f"{_BASE_SNAP_DIR}/current"
-    _SNAP_COMMON = f"{_BASE_SNAP_DIR}/common"
-    _SNAP = "/snap/opensearch/current"
+    _BASE_SNAP_DIR = BASE_SNAP_DIR
+    _SNAP_DATA = SNAP_DATA
+    _SNAP_COMMON = SNAP_COMMON
+    _SNAP = SNAP
 
     def __init__(self, charm, peer_relation: str):
+        for attempt in Retrying(stop=stop_after_attempt(5), wait=wait_fixed(wait=5)):
+            with attempt:
+                cache = snap.SnapCache()
+                self._opensearch = cache["opensearch"]
         super().__init__(charm, peer_relation)
 
+    def _update_snap_metadata(self):
```

Comment:
i don't think we should have this - if we only support upgrades from a version with upgrades (the current 2.12)

Reply:
Upgrade from 2.11.1 > 2.12.0 also moves from JDK v17 > JDK v21. That is a move in the environment variables. 

However, the upgrade works as follow:
1. Every unit refreshes its own charm (hence, downloads the new charm code)
2. The unit waits to be allowed to upgrade its snap (hence, running an older version of the snap meanwhile)
3. That may spans over several hooks (hence, several creations of the `OpenSearchDistribution` object, which creates the environment variables).

I did not add it because I want, but because a simple 2.11.1 > 2.12.0 (as tested in the new integration test) breaks otherwise.

Other examples, within a minor revision we had the update of folder structures or even migrating to openjdk.

Reply:
we only support upgrades from `2.12.0` -> `2.13.0` and up from there

Reply:
@Mehdi-Bendriss it is not about the version, but that will come up once again. We've changed the behavior of `Paths` at least 3x times.